src/NServiceBus.Core/Pipeline/Outgoing/OutgoingContext.cs

```diff
@@ -11,10 +11,14 @@ protected OutgoingContext(string messageId, Dictionary<string, string> headers,
             : base(parentContext)
         {
             Headers = headers;
-            MessageId = messageId;
+
+            if (!Headers.ContainsKey(NServiceBus.Headers.MessageId))
```

Comment:
not entirely convinced by this approach. Why not do this on `ExtendableOptions` or set the header when we create the `OutgoingSendContext` (+the related contexts)? In our discussion we agreed that keeping `MessageId` isn't even a requirement? I think I'm still on that side. If you decide to mess with headers, don't expect other properties to automatically update, we don't do that with other context properties.